enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,356 @@
+  replicating the functionality of the `main()` function.
+- A duplicate of GDP's configuration structs may be used in MicroShift's 
+  configuration code to make it more readable if needed.
+- GDP will be an optional feature, disabled by default, but included within
+  the MicroShift binary.
+- MicroShift will expose the devices with `device.microshift.io/NAME` label
+  (changed via configuration from `squat.ai/NAME`).
+
+### Risks and Mitigations
+
+- Making changes to the fork may result in an additional effort when syncing
+  upstream changes (e.g. resolving conflicts; unless we contribute the changes).
+  Fortunately, GDP is not very complex product, nor it's subject to fast paced
+  development because for its purpose it's mostly complete.
+
+### Drawbacks
+
+MicroShift team will need to manage the go.mod importing with all the
+consequences of the Go modules versioning.
+
+## Alternatives (Not Implemented)
+
+### Container approach
+
+An alternative to importing the code directly is the "container" approach.
+It involves publishing the forked Generic Device Plugin as an container image
+along with a manifest for MicroShift.
+This might align better with Kubernetes principles (containerization) and could
+offer broader reuse (but not necessarily by OpenShift right now), it presents
+several drawbacks:
+- Increased maintenance effort on MicroShift team:
+  As GDP would likely not be a part of the OpenShift payload and MicroShift team
+  gets help from ART in managing the RPMs. The MicroShift team does not have
+  experience in productizing images.
+- Size of the Generic Device Plugin image would likely be significantly larger
+  than increase in the MicroShift binary size from code import. This is
+  important consideration due to MicroShift's target environment.
+
+### Other device plugin solution
+
+Investigating this feature only three solutions were found:
+- Generic Device Plugin (discussed in this enhancement)
+- [Smarter Device Plugin](https://gitlab.com/arm-research/smarter/smarter-device-manager) (abandoned, last commit in Sept 2022)
+- [Akri](https://docs.akri.sh/)
+
+The Smarter Device Plugin was not considered as it is no longer actively maintained.
+
+Akri, while an interesting approach that abstracts the Device Plugin framework,
+was deemed too complex for our requirements. The MicroShift team concluded that
+integrating Akri would significantly increase maintenance effort across
+engineering, QE, and documentation.
+
+Since the Generic Device Plugin (GDP) can be disabled via configuration, users
+retain the flexibility to utilize upstream Akri if their needs require it.
+
+## Open Questions [optional]
+
+## Test Plan
+
+Usually device plugins are tested using real hardware to make sure everything works.
+
+We'll need to get creative and implement some kind of dummy serial device to
+test communication between Pod utilizing the device and fake device
+(in the form of a daemon on the system).
```

Comment:
There should be a device simulator we can use for such tests
See https://github.com/cminyard/serialsim

Reply:
Yeah, there's also a `socat` which is shipped in RHEL:
```
$ sudo socat -d -d PTY,link=/dev/faketty0 PTY,link=/dev/fakettyS0
```